sys/net/netstats/netstats_neighbor.c

```diff
@@ -0,0 +1,410 @@
+    }
+
+#if MODULE_NETSTATS_NEIGHBOR_TX_TIME
+    stats->time_tx_avg = _ewma(fresh, stats->time_tx_avg, duration);
+#else
+    (void)stats;
+    (void)result;
+    (void)duration;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_update_rssi(netstats_nb_t *stats, uint8_t rssi, bool fresh)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_RSSI
+    stats->rssi = _ewma(fresh, stats->rssi, rssi);
+#else
+    (void)stats;
+    (void)rssi;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_update_lqi(netstats_nb_t *stats, uint8_t lqi, bool fresh)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_LQI
+    stats->lqi = _ewma(fresh, stats->lqi, lqi);
+#else
+    (void)stats;
+    (void)lqi;
+    (void)fresh;
+#endif
+}
+
+static void netstats_nb_incr_count_tx(netstats_nb_t *stats, netstats_nb_result_t result)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_COUNT
+    stats->tx_count++;
+
+    /* gracefully handle overflow */
+    if (stats->tx_count == 0) {
+        stats->tx_count = ~stats->tx_count;
+        stats->tx_count = stats->tx_count >> 4;
+        stats->tx_fail  = stats->tx_fail  >> 4;
+    }
+
+    if (result != NETSTATS_NB_SUCCESS) {
+        stats->tx_fail++;
+    }
+#else
+    (void)stats;
+    (void)result;
+#endif
+}
+
+static void netstats_nb_incr_count_rx(netstats_nb_t *stats)
+{
+#ifdef MODULE_NETSTATS_NEIGHBOR_COUNT
+    stats->rx_count++;
+#else
+    (void)stats;
+#endif
+}
+
+netstats_nb_t *netstats_nb_update_tx(netif_t *dev, netstats_nb_result_t result, uint8_t transmissions)
+{
+    uint32_t now = xtimer_now_usec();
+    netstats_nb_t *stats;
+    uint32_t time_tx = 0;
+
+    _lock(dev);
+
+    /* Buggy drivers don't always generate TX done events.
+     * Discard old events to prevent the tx start <-> tx done correlation
+     * from getting out of sync. */
+    do {
+        stats = netstats_nb_get_recorded(dev, &time_tx);
+    } while (cib_avail(&dev->neighbors.stats_idx) &&
+            ((now - time_tx) > NETSTATS_NB_TX_TIMEOUT_MS * US_PER_MS));
+
+    /* Nothing to do for multicast or if packet was not sent */
+    if (result == NETSTATS_NB_BUSY || stats == NULL) {
+        goto out;
+    }
+
+    bool fresh = isfresh(stats);
+
+    netstats_nb_update_time(stats, result, now - time_tx, fresh);
+    netstats_nb_update_etx(stats, result, transmissions, fresh);
+    netstats_nb_incr_count_tx(stats, result);
+
+    incr_freshness(stats);
+
+out:
+    _unlock(dev);
+    return stats;
+}
+
+netstats_nb_t *netstats_nb_update_rx(netif_t *dev, const uint8_t *l2_addr,
+                                     uint8_t l2_addr_len, uint8_t rssi, uint8_t lqi)
+{
+    _lock(dev);
+
+    netstats_nb_t *stats = netstats_nb_get_or_create(dev, l2_addr, l2_addr_len);
+
+    if (stats == NULL) {
+        goto out;
+    }
+
+    bool fresh = isfresh(stats);
+
+    netstats_nb_update_rssi(stats, rssi, fresh);
+    netstats_nb_update_lqi(stats, lqi, fresh);
+    netstats_nb_incr_count_rx(stats);
+
+    incr_freshness(stats);
+
+out:
+    _unlock(dev);
+    return stats;
```

Comment:
```suggestion
    if (stats != NULL) {
        bool fresh = isfresh(stats);

        netstats_nb_update_rssi(stats, rssi, fresh);
        netstats_nb_update_lqi(stats, lqi, fresh);
        netstats_nb_incr_count_rx(stats);

        incr_freshness(stats);
    }
    _unlock(dev);
    return stats;
```

And the `goto` and the stray variable declaration disappear ;-)